src/lib/transition_frontier/transition_frontier.ml

```diff
@@ -857,43 +863,52 @@ struct
-                  ~expect:(Ledger_proof.statement proof_data).source
-                  ( Ledger.Db.merkle_root t.root_snarked_ledger
-                  |> Frozen_ledger_hash.of_ledger_hash ) ;
-                (* Apply all the transactions associated with the new ledger
-                   proof to the database-backed SNARKed ledger. We create a
-                   mask and apply them to that, then commit it to the DB. This
-                   saves a lot of IO since committing is batched. Would be even
-                   faster if we implemented #2760. *)
-                let db_casted =
-                  Ledger.Any_ledger.cast
-                    (module Ledger.Db)
-                    t.root_snarked_ledger
-                in
-                let db_mask =
-                  Ledger.Maskable.register_mask db_casted
-                    (Ledger.Mask.create ())
-                in
-                Non_empty_list.iter txns ~f:(fun txn ->
-                    Ledger.apply_transaction db_mask txn
-                    |> Or_error.ok_exn |> ignore ) ;
-                Ledger.commit db_mask ;
-                ignore @@ Ledger.Maskable.unregister_mask_exn db_casted db_mask
-            | _, false | None, _ ->
-                () ) ;
+              | Some txns, true ->
+                  let proof_data =
+                    Staged_ledger.current_ledger_proof new_root_staged_ledger
+                    |> Option.value_exn
+                  in
+                  [%test_result: Frozen_ledger_hash.t]
+                    ~message:
+                      "Root snarked ledger hash should be the same as the \
+                       source hash in the proof that was just emitted"
+                    ~expect:(Ledger_proof.statement proof_data).source
+                    ( Ledger.Db.merkle_root t.root_snarked_ledger
+                    |> Frozen_ledger_hash.of_ledger_hash ) ;
+                  (* Apply all the transactions associated with the new ledger
+                     proof to the database-backed SNARKed ledger. We create a
+                     mask and apply them to that, then commit it to the DB. This
+                     saves a lot of IO since committing is batched. Would be even
+                     faster if we implemented #2760. *)
+                  let db_casted =
+                    Ledger.Any_ledger.cast
+                      (module Ledger.Db)
+                      t.root_snarked_ledger
+                  in
+                  let db_mask =
+                    Ledger.Maskable.register_mask db_casted
+                      (Ledger.Mask.create ())
+                  in
+                  let%bind () =
+                    Non_empty_list.iter_deferred txns ~f:(fun txn ->
+                        let%map () = Async.Scheduler.yield () in
+                        ignore
+                          ( Ledger.apply_transaction db_mask txn
+                          |> Or_error.ok_exn ) )
```

Comment:
Hm, this one specifically can probably be a Deferred.return? The behavior of iter_deferred should have the yield points you want.

Reply:
True, I'll just need to add a yield before the first iteration.